### PR TITLE
Deprecate CLI factory in azure-common

### PR DESCRIPTION
**Pull request by lmazuel**
See https://github.com/Azure/azure-sdk-for-python/pull/20657 and https://github.com/Azure/azure-sdk-for-python/issues/21313

### CODE REVIEW

**Comment by lmazuel** [comment]
image

File: sdk/core/azure-common/azure/common/credentials.py [hunk 1]
```diff
@@ -27,7 +30,10 @@ def get_cli_profile():
         from azure.cli.core._session import ACCOUNT
         from azure.cli.core._environment import get_config_dir
     except ImportError:
-        raise ImportError("You need to install 'azure-cli-core' to load CLI credentials")
+        raise ImportError(
+            "You need to install 'azure-cli-core<2.21.0' to load CLI credentials using this method. " +
```
[the remark this answers is not in the document]
Comment (lmazuel) [reply sdk/core/azure-common/azure/common/credentials.py]: Redone it

File: sdk/core/azure-common/azure/common/client_factory.py [hunk 2]
```diff
@@ -50,7 +51,18 @@ def _client_resource(client_class, cloud):
...
+    *Disclaimer*: This is NOT the recommended approach to authenticate with CLI login, this method is deprecated.
+    use https://pypi.org/project/azure-identity/ and AzureCliCredential instead. See example code below:
+
+    .. code:: python
+
+        from azure.identity import AzureCliCredential
+        from azure.mgmt.compute import ComputeManagementClient
+        client = ComputeManagementClient(AzureCliCredential(), subscription_id)
+
+    This method is not working for CLI installation after 3/2021 (version 2.21.0 of azure-cli-core).
```
Comment (annatisch) [inline sdk/core/azure-common/azure/common/client_factory.py]: This method will not work for CLI installations after 3/2021

File: sdk/core/azure-common/azure/common/client_factory.py [hunk 3]
```diff
@@ -120,11 +142,7 @@ def _is_autorest_v3(client_class):
 def get_client_from_json_dict(client_class, config_dict, **kwargs):
     """Return a SDK client initialized with a JSON auth dict.
 
-    The easiest way to obtain this content is to call the following CLI commands:
-
-    .. code:: bash
-
-        az ad sp create-for-rbac --sdk-auth
+    *Disclaimer*: This is NOT recommended approach, see https://aka.ms/azsdk/python/identity/migration for guidance.
```
Comment (annatisch) [inline sdk/core/azure-common/azure/common/client_factory.py:145]: "the" recommended approach

File: sdk/core/azure-common/azure/common/credentials.py [hunk 4]
```diff
@@ -81,19 +88,51 @@ def signed_session(self, session=None):
 def get_azure_cli_credentials(resource=None, with_tenant=False):
     """Return Credentials and default SubscriptionID of current loaded profile of the CLI.
 
-    Credentials will be the "az login" command:
+    *Disclaimer*: This method is not working for CLI installation after 3/2021 (version 2.21.0 of azure-cli-core).
```
[the remark this answers is not in the document]
Comment (lmazuel) [reply sdk/core/azure-common/azure/common/credentials.py:94]: pushing now

File: sdk/core/azure-common/azure/common/cloud.py [hunk 5]
```diff
@@ -17,6 +21,10 @@ def get_cli_active_cloud():
     try:
         from azure.cli.core.cloud import get_active_cloud
     except ImportError:
-        raise ImportError("You need to install 'azure-cli-core' to load CLI active Cloud")
+        raise ImportError(
+            "Public API of azure-cli-core have been deprecated starting 2.21.0, " +
```
Comment (annatisch) [inline sdk/core/azure-common/azure/common/cloud.py:25]: "The public API..."
"has been deprecated"

File: sdk/core/azure-common/azure/common/credentials.py [hunk 6]
```diff
@@ -27,8 +32,12 @@ def get_cli_profile():
         from azure.cli.core._session import ACCOUNT
         from azure.cli.core._environment import get_config_dir
     except ImportError:
-        raise ImportError("You need to install 'azure-cli-core' to load CLI credentials")
-
+        raise ImportError(
+            "Public API of azure-cli-core have been deprecated starting 2.21.0, " +
+            "and this method no longer can return a profile. " +
```
Comment (annatisch) [inline sdk/core/azure-common/azure/common/credentials.py:37]: "can no longer return"

File: sdk/core/azure-common/azure/common/credentials.py [hunk 7]
```diff
@@ -81,19 +91,58 @@ def signed_session(self, session=None):
 def get_azure_cli_credentials(resource=None, with_tenant=False):
     """Return Credentials and default SubscriptionID of current loaded profile of the CLI.
 
-    Credentials will be the "az login" command:
+    *Disclaimer*: This method is not working for azure-cli-core>=2.21.0 (released in March 2021).
+    Recommended authentication is now to use https://pypi.org/project/azure-identity/ and AzureCliCredential.
```
Comment (annatisch) [inline sdk/core/azure-common/azure/common/credentials.py:95]: "It is now recommended to authenticate using..."

File: sdk/core/azure-common/azure/common/credentials.py [hunk 8]
```diff
@@ -81,19 +88,51 @@ def signed_session(self, session=None):
...
 
+    .. deprecated:: 1.1.28
+
     :param str resource: The alternative resource for credentials if not ARM (GraphRBac, etc.)
     :param bool with_tenant: If True, return a three-tuple with last as tenant ID
     :return: tuple of Credentials and SubscriptionID (and tenant ID if with_tenant)
     :rtype: tuple
     """
+    warnings.warn(
+        "get_client_from_cli_profile is deprecated, please use azure-identity and AzureCliCredential instead",
```
[the remark this answers is not in the document]
Comment (lmazuel) [reply sdk/core/azure-common/azure/common/credentials.py]: added a aka.ms link

File: sdk/core/azure-common/azure/common/client_factory.py (hunk 2 above)
Comment (lmazuel) [reply sdk/core/azure-common/azure/common/client_factory.py]: changed the complete sentence after a comment from Johan on ambiguous guidance between dates and version

**Review by annatisch — APPROVED** [review]

**Comment by xiangyan99** [comment]
None of the failures is related to the change.

**Pull request merged**